compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2714,6 +2718,27 @@ impl<'a> Parser<'a> {
         if let ExprKind::Let(_, _, _, Recovered::No) = cond.kind {
             // Remove the last feature gating of a `let` expression since it's stable.
             self.psess.gated_spans.ungate_last(sym::let_chains, cond.span);
+        } else {
+            fn ungate_let_exprs(this: &mut Parser<'_>, expr: &Expr) {
+                if !expr.span.at_least_rust_2024() {
+                    return;
+                }
+                match &expr.kind {
+                    ExprKind::Binary(BinOp { node: BinOpKind::And, .. }, lhs, rhs) => {
+                        ungate_let_exprs(this, rhs);
+                        ungate_let_exprs(this, lhs);
+                    }
+                    ExprKind::Let(..) => {
+                        this.psess.gated_spans.ungate_last(sym::let_chains, expr.span)
```

Comment:
The docs for this function say:

> Using this is discouraged unless you have a really good reason to.

Perhaps we should add a comment to document our good reason.

Also, since it seems unlikely to me that we'll ever stabilize let chains in Rust `..=2021` -- that is, unless we ever decide to make a breaking change to earlier editions to align the drop order rules -- I wonder if it wouldn't be better to just pull let chains out of older editions entirely and remove use of the feature gate.

Reply:
> I wonder if it wouldn't be better to just pull let chains out of older editions entirely and remove use of the feature gate.

Back when I wrote the PR, the compiler has still been on edition 2021, and 2024 hasn't been stable yet. So it was a hard criterion to support 2021 in some way, too.

Now it's months later, and I guess it's feasible also to nightly users that we can immediately drop support for edition 2021 and below. But I'd still prefer to do that in a followup to this PR.

Reply:
> Perhaps we should add a comment to document our good reason.

I wanted to be careful, especially when seeing PRs like #95008 and #98633. But I think with #115677, we probably have good recognition of when some `let` is in a disallowed location. So I've made a change to remove `ungate_last` usages and to instead not gate in the first place.